File: 卡码/代码随想录/1/09-动态规划/Go/01-基础题目/LeetCode_0343.go
// 343. 整数拆分
package main

import "math"

func integerBreak(n int) int {
	dp:=make([]int,n+1)
	for i:=0;i<len(dp);i++ {
		dp[i]=0
	}
	dp[0],dp[1],dp[2]=0,0,1
	for i:=3;i<n+1;i++ {
		for j:=1;j<=i/2;j++ {
			dp[i]=int(math.Max(float64(dp[i]),math.Max(float64(j*(i-j)),float64(j*dp[i-j]))))
		}
	}
	return dp[n]
}

File: 卡码/代码随想录/1/09-动态规划/Python/01-基础题目/LeetCode_0343.py
#343. 整数拆分
from typing import *

class Solution:
  def integerBreak(self, n: int) -> int:
    dp:List[int]=[0]*(n+1)
    dp[0],dp[1],dp[2]=0,0,1
    for i in range(3,n+1):
      for j in range(1,i//2+1):
        dp[i]=max(dp[i],max(j*(i-j),j*dp[i-j]))
    return dp[n]